### ISSUE TITLE
Cookie expiration time is not extended when browsing the frontend

### ISSUE THREAD

**Issue by xi-ao** [magento/magento2#7627]
Unlike M1, session cookie expiration date is never extended with each page the user visits. Which means that after logging in, if the session lifetime is set to 20 minutes for example, the user will be disconnected after this time even if he spent the whole time browsing the pages on the frontend.

Tested on a blank M2 v2.1.2 with sample data.

### Preconditions
1. Install M2 with sample data

### Steps to reproduce
1. Go to frontend
2. Create an account and stay connected
3. Look at your PHPSESSID and its expiraton date in the browser's console tab
4. Browse pages of the catalog
5. Check the cookie expiration time
6. (repeat)

### Expected result
With each page visited, expiration date of the PHPSESSID cookie should be set to "now + session lifetime"

### Actual result
PHPSESSID expiration date is never updated, and the user is logged out after the session lifetime.
Only opening the cart page regenerates the cookie (not only updating but **regenerating**) and allows to browse longer than than the session lifetime.
(*sometimes* adding items to cart does too)

**Comment by srirammoorthyziffity** [comment]
I am facing the same issue. Is there any update?

**Comment by srirammoorthyziffity** [comment]
@veloraven I could see this bug exist in 2.0 as well please update.

**Comment by mrfroasty** [comment]
Any news this  bug was reported in november last year .... bump :d

**Comment by EugeneAinstainer** [comment]
Same issue. Is there any update?

**Comment by sma09sjb** [comment]
This needs to be fixed.

**Comment by andreaskoch** [comment]
I have the same issue in Magento Enterprise Edition 2.1.2 and it is not only related to frontend pages. The expiration date of the PHPSESSID cookie is also not updated on the cart page.

Once you go to the checkout the expiration date of the PHPSESSID cookie is updated. But until a customer reaches the checkout the **session expiration effectively acts as a countdown** that he cannot escape. Once time is up. The cart will be gone.

I can reproduce the same behaviour fresh installation of Magento 2.1.5 CE with just sample data installed.

**Comment by xi-ao** [comment]
One way to temporarily fix this is to create an observer on `controller_action_predispatch` on the frontend and use the method `\Magento\Framework\Stdlib\Cookie\PhpCookieManager::setPublicCookie()` to renew the cookie. It takes 2 minutes and should solve the issue.

**Comment by dnadle** [comment]
Any update on this? It's still happening in 2.1.6.

**Comment by vityakopin** [comment]
Internal ticket MAGETWO-69958 was created.

**Comment by damcou** [comment]
Hi!
Is there any news about this bug ?
I've tried the xi-ao's solution (the controller action's observer). The cookie expiration is well renewed as I want but after, it's impossible to login as a customer on the frontend. 

This is my code in the execute function, Am I doing someting wrong ?

```
    public function execute(\Magento\Framework\Event\Observer $observer)
    {
        $cookie = ($this->cookieManager->getCookie('PHPSESSID'));
        if ($cookie) {
            $publicCookieMetadata = $this->cookieMetadataFactory->createPublicCookieMetadata()
                ->setDuration(3600)
                ->setPath($this->sessionConfig->getCookiePath())
                ->setDomain($this->sessionConfig->getCookieDomain())
                ->setSecure(false)
                ->setHttpOnly(true);

            $this->cookieManager->setPublicCookie('PHPSESSID', $cookie, $publicCookieMetadata);
        }
    }
```

**Comment by xi-ao** [comment]
@Dam62 This is the code I use for my plugin, hope it helps:

```php
class SessionCookieRefresh Implements \Magento\Framework\Event\ObserverInterface
{
    /**
     * @var \Magento\Customer\Model\Session
     */
    protected $session;

    /**
     * @var \Magento\Framework\Stdlib\Cookie\PhpCookieManager
     */
    protected $cookieManager;

    /**
     * @var \Magento\Framework\Stdlib\Cookie\CookieMetadataFactory
     */
    protected $cookieMetadataFactory;

    /**
     * @var \Magento\Framework\Session\Config\ConfigInterface
     */
    protected $sessionConfig;

    /**
     * SessionCookieRefresh constructor.
     * @param \Magento\Customer\Model\Session $session
     * @param \Magento\Framework\Session\Config\ConfigInterface $sessionConfig
     * @param \Magento\Framework\Stdlib\Cookie\CookieMetadataFactory $cookieMetadataFactory
     * @param \Magento\Framework\Stdlib\Cookie\PhpCookieManager $cookieManager
     */
    public function __construct(
        \Magento\Customer\Model\Session $session,
        \Magento\Framework\Session\Config\ConfigInterface $sessionConfig,
        \Magento\Framework\Stdlib\Cookie\CookieMetadataFactory $cookieMetadataFactory,
        \Magento\Framework\Stdlib\Cookie\PhpCookieManager $cookieManager
    ) {
        $this->session = $session;
        $this->sessionConfig = $sessionConfig;
        $this->cookieMetadataFactory = $cookieMetadataFactory;
        $this->cookieManager = $cookieManager;
    }

    /**
     * @param \Magento\Framework\Event\Observer $observer
     *
     * refreshes de duration of the session cookie
     */
    public function execute(\Magento\Framework\Event\Observer $observer)
    {
        $metadata = $this->cookieMetadataFactory->createPublicCookieMetadata();
        $metadata->setPath($this->session->getCookiePath())
            ->setDomain($this->session->getCookieDomain())
            ->setDuration($this->session->getCookieLifetime())
            ->setSecure($this->sessionConfig->getCookieSecure())
            ->setHttpOnly($this->sessionConfig->getCookieHttpOnly());

        $this->cookieManager->setPublicCookie(
            $this->session->getName(),
            $this->session->getSessionId(),
            $metadata
        );
    }
}
```

**Comment by damcou** [comment]
@xi-ao  Thank you very much for your quick answer ! 
I checked the differences between our versions then I fixed mine (I wasn't using customer session model as you probably noticed).

It still not works on Chrome but works perfectly on Firefox so I think the problem comes from my computer for a reason or another.

Thanks again !

**Comment by jbokhari** [comment]
The solution @xi-ao posted worked, however I started having trouble logging in with the solution in place. Clearing cookies and using incognito window did not help. I tried several other fixes online to no avail and had to finally revert the changes. Still looking into issue to find a better solution.

Also wanted to add that I'm getting this issue in CE 2.1.7. I did notice that the checkout pages may be refreshing the cookie, but I haven't verified this. Regardless that doesn't help much.

**Comment by magento-engcom-team** [comment]
@xi-ao, thank you for your report.
We've created internal ticket(s) MAGETWO-69958 to track progress on the issue.

**Comment by arlycreative** [comment]
Is this issue solved yet?

**Comment by tlygnersjo** [comment]
Is this issue solved yet?

**Comment by p-bystritsky** [comment]
@tlygnersjo no.

**Comment by arnoudhgz** [comment]
@magento-engcom-team @sidolov  Am I correct that this is solved in (at least) 2.2.5? I don't have this issue, when browsing the catalog pages the cookie gets updated properly.

**Comment by magento-engcom-team** [comment]
Unfortunately, we are archiving this ticket now as it did not get much attention from both Magento Community and Core developers for an extended period. This is done in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to comment or reopen according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
the ticket if you are still facing this issue on the latest `2.x-develop` branch. Thank you for collaboration.
